// order/test/order.js
(function() {
  var Order;

  Order = require('../order');

  module.exports = {
    async: function(æ) {
      var list, n, results;
      n = 0;
      results = [
        {
          idx: 1,
          before: -1,
          after: -1
        }, {
          idx: 2,
          before: +1,
          after: -1
        }, {
          idx: 3,
          before: +2,
          after: -1
        }, {
          idx: 0,
          before: -1,
          after: +1
        }
      ];
      list = new Order(function(e) {
        return æ.deepEqual(e, results.shift());
      });
      list.push((function(done) {
        setTimeout(done, 30);
        return n++;
      }));
      list.push((function(done) {
        setTimeout(done, 10);
        return n++;
      }));
      list.push((function(done) {
        setTimeout(done, 20);
        return n++;
      }));
      list.insert(1, (function(done) {
        setTimeout(done, 4);
        return n++;
      }));
      æ.equal(list.length, 4);
      æ.deepEqual(list.slice(), [0, 3, 1, 2]);
      æ.deepEqual(list.done, [false, false, false, false]);
      return setTimeout(function() {
        æ.equal(results.length, 0);
        æ.deepEqual(list.done, [true, true, true, true]);
        return æ.done();
      }, 32);
    },
    sync: function(æ) {
      var list, n, results;
      n = 0;
      results = [
        {
          idx: 0,
          before: -1,
          after: -1
        }, {
          idx: 1,
          before: +0,
          after: -1
        }, {
          idx: 2,
          before: +1,
          after: -1
        }, {
          idx: 1,
          before: +0,
          after: +2
        }
      ];
      list = new Order(function(e) {
        return æ.deepEqual(e, results.shift());
      });
      list.push(function(done) {
        done();
        return n++;
      });
      list.push(function(done) {
        done();
        return n++;
      });
      list.push(function(done) {
        done();
        return n++;
      });
      list.insert(1, function(done) {
        done();
        return n++;
      });
      æ.equal(list.length, 4);
      æ.deepEqual(list.slice(), [0, 3, 1, 2]);
      æ.deepEqual(list.done, [true, true, true, true]);
      æ.equal(results.length, 0);
      return æ.done();
    },
    'get entry when sync done': function(æ) {
      var list, n, results;
      n = 0;
      results = [
        [
          {
            idx: 0,
            before: -1,
            after: -1
          }, 0
        ], [
          {
            idx: 1,
            before: +0,
            after: -1
          }, 1
        ]
      ];
      list = new Order(function(e) {
        var r;
        r = results.shift();
        æ.deepEqual(e, r[0]);
        return æ.equal(this[e.idx], r[1]);
      });
      list.push(function(done) {
        done();
        return n++;
      });
      list.push(function(done) {
        done();
        return n++;
      });
      æ.equal(list.length, 2);
      æ.deepEqual(list.slice(), [0, 1]);
      æ.deepEqual(list.done, [true, true]);
      æ.equal(results.length, 0);
      return æ.done();
    },
    rm: function(æ) {
      var list, n, results;
      n = 0;
      results = [
        {
          idx: 0,
          before: -1,
          after: -1
        }, {
          idx: 1,
          before: +0,
          after: -1
        }, {
          idx: 2,
          before: +1,
          after: -1
        }, {
          idx: 3,
          before: +2,
          after: -1
        }
      ];
      list = new Order(function(e) {
        return æ.deepEqual(e, results.shift());
      });
      list.push(function(done) {
        done();
        return n++;
      });
      list.push(function(done) {
        done();
        return n++;
      });
      list.push(function(done) {
        done();
        return n++;
      });
      list.push(function(done) {
        done();
        return n++;
      });
      list.remove(1);
      æ.deepEqual(list.slice(), [0, 2, 3]);
      æ.equal(list.done.length, 3);
      æ.equal(list.keys.length, 3);
      æ.equal(list.length, 3);
      list.pop();
      æ.deepEqual(list.slice(), [0, 2]);
      æ.equal(list.done.length, 2);
      æ.equal(list.keys.length, 2);
      æ.equal(list.length, 2);
      list.shift();
      æ.deepEqual(list.slice(), [2]);
      æ.equal(list.done.length, 1);
      æ.equal(list.keys.length, 1);
      æ.equal(list.length, 1);
      æ.equal(results.length, 0);
      return æ.done();
    },
    'no double done': function(æ) {
      var list, n, results;
      n = 0;
      results = [
        {
          idx: 1,
          before: -1,
          after: -1
        }, {
          idx: 2,
          before: +1,
          after: -1
        }, {
          idx: 3,
          before: +2,
          after: -1
        }, {
          idx: 0,
          before: -1,
          after: +1
        }
      ];
      list = new Order(function(e) {
        return æ.deepEqual(e, results.shift());
      });
      list.push((function(done) {
        setTimeout(done, 30);
        return n++;
      }));
      list.push((function(done) {
        setTimeout(done, 10);
        setTimeout(done, 23);
        return n++;
      }));
      list.push((function(done) {
        setTimeout(done, 20);
        return n++;
      }));
      list.insert(1, (function(done) {
        setTimeout(done, 4);
        return n++;
      }));
      æ.equal(list.length, 4);
      æ.deepEqual(list.slice(), [0, 3, 1, 2]);
      æ.deepEqual(list.done, [false, false, false, false]);
      return setTimeout(function() {
        æ.equal(results.length, 0);
        æ.deepEqual(list.done, [true, true, true, true]);
        return æ.done();
      }, 32);
    },
    splice: function(æ) {
      var l, list, n, results;
      n = 0;
      results = [
        {
          idx: 3,
          before: -1,
          after: -1
        }, {
          idx: 0,
          before: -1,
          after: +3
        }, {
          idx: 1,
          before: +0,
          after: +3
        }, {
          idx: 2,
          before: +1,
          after: +3
        }
      ];
      list = new Order(function(e) {
        return æ.deepEqual(e, results.shift());
      });
      list.push((function(done) {
        var a;
        setTimeout((function() {
          a[0] = n++;
          return done();
        }), 30);
        return a = [n++];
      }));
      list.push((function(done) {
        var a;
        setTimeout((function() {
          a[0] = n++;
          return done();
        }), 20);
        return a = [n++];
      }));
      list.push((function(done) {
        var a;
        setTimeout((function() {
          a[0] = n++;
          return done();
        }), 10);
        return a = [n++];
      }));
      æ.equal(list.length, 3);
      æ.deepEqual((function() {
        var _i, _len, _results;
        _results = [];
        for (_i = 0, _len = list.length; _i < _len; _i++) {
          l = list[_i];
          _results.push(l.slice());
        }
        return _results;
      })(), [[0], [1], [2]]);
      æ.deepEqual(list.done, [false, false, false]);
      list.splice(1, 1, (function(done) {
        var a;
        setTimeout((function() {
          a[0] = n++;
          return done();
        }), 40);
        return a = [n++];
      }), (function(done) {
        var a;
        setTimeout((function() {
          a[0] = n++;
          return done();
        }), 50);
        return a = [n++];
      }));
      æ.equal(list.length, 4);
      æ.deepEqual((function() {
        var _i, _len, _results;
        _results = [];
        for (_i = 0, _len = list.length; _i < _len; _i++) {
          l = list[_i];
          _results.push(l.slice());
        }
        return _results;
      })(), [[0], [3], [4], [2]]);
      æ.deepEqual(list.done, [false, false, false, false]);
      return setTimeout(function() {
        var l;
        æ.equal(results.length, 0);
        æ.deepEqual((function() {
          var _i, _len, _results;
          _results = [];
          for (_i = 0, _len = list.length; _i < _len; _i++) {
            l = list[_i];
            _results.push(l.slice());
          }
          return _results;
        })(), [[7], [8], [9], [5]]);
        æ.deepEqual(list.done, [true, true, true, true]);
        return æ.done();
      }, 52);
    }
  };

}).call(this);
